### PR TITLE
Run waitpid on childs to ensure no zombies are left behind

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
We can call waitpid on the process to reparent and ensure no zombies are left behind.
#### Which issue(s) this PR fixes:

Fixes https://github.com/containers/crun/issues/1482
Fixes https://github.com/containers/conmon-rs/issues/2242

#### Special notes for your reviewer:
None
#### Does this PR introduce a user-facing change?

```release-note
Fixed issue that containers cannot be removed if init command was not found.
```

### CODE REVIEW

**Comment by saschagrunert** [comment]
Fixes https://github.com/containers/conmon-rs/issues/2242

File: conmon-rs/server/src/child_reaper.rs [hunk 1]
```diff
@@ -137,6 +137,9 @@ impl ChildReaper {
             // meaning there is no other entity that could cancel the read_loops.
             token.cancel();
 
+            // Ensure to wait for all childs to not become zombies.
```
Comment (kwilczynski) [inline conmon-rs/server/src/child_reaper.rs:140]:
```suggestion
            // Wait to ensure that children do not become zombies.
```

File: conmon-rs/server/src/child_reaper.rs [hunk 2]
```diff
@@ -149,6 +152,34 @@ impl ChildReaper {
...
+        debug!("Checking child processes");
+        let pid = Pid::from_raw(-1);
+        loop {
+            match waitpid(pid, WaitPidFlag::WNOHANG.into()) {
+                Ok(WaitStatus::Exited(p, code)) => {
+                    debug!("PID {p} exited with status: {code}");
+                    break;
+                }
+                Ok(WaitStatus::StillAlive) => {
+                    debug!("Process is still in same state, breaking");
```
Comment (kwilczynski) [inline conmon-rs/server/src/child_reaper.rs:165]:
```suggestion
                    debug!("PID {p} is still in same state");
```

File: conmon-rs/server/src/child_reaper.rs [hunk 3]
```diff
@@ -149,6 +152,34 @@ impl ChildReaper {
...
+                    break;
+                }
+                Ok(_) => {
+                    continue;
+                }
+                Err(Errno::EINTR) => {
+                    debug!("Failed to wait for PID on EINTR, retrying");
```
Comment (kwilczynski) [inline conmon-rs/server/src/child_reaper.rs:172]:
```suggestion
                    debug!("Retrying on EINTR for PID {pid}");
```

Comment (saschagrunert) [reply conmon-rs/server/src/child_reaper.rs]: Fixed as suggested :+1:

File: conmon-rs/server/src/child_reaper.rs (hunk 2 above)
Comment (saschagrunert) [reply conmon-rs/server/src/child_reaper.rs]: Fixed as suggested :+1:

File: conmon-rs/server/src/child_reaper.rs (hunk 1 above)
Comment (saschagrunert) [reply conmon-rs/server/src/child_reaper.rs]: Fixed as suggested :+1:

File: conmon-rs/server/src/child_reaper.rs [hunk 4]
```diff
@@ -149,6 +152,34 @@ impl ChildReaper {
...
+    fn check_child_processes() {
+        debug!("Checking child processes");
+        let pid = Pid::from_raw(-1);
+        loop {
+            match waitpid(pid, WaitPidFlag::WNOHANG.into()) {
+                Ok(WaitStatus::Exited(p, code)) => {
+                    debug!("PID {p} exited with status: {code}");
+                    break;
+                }
+                Ok(WaitStatus::StillAlive) => {
```
Comment (kwilczynski) [inline conmon-rs/server/src/child_reaper.rs:164]: Do we also want to match here on  `WaitStatus::Stopped()` and `WaitStatus::Continued()`? I am thinking about a case when the process suddenly _goes away_ for reasons outside of conmon-rs' influence.

Comment (kwilczynski) [reply conmon-rs/server/src/child_reaper.rs:164]: On the other hand, perhaps leaving it `Ok()` is fine, to skip over, so to speak.

Comment (kwilczynski) [reply conmon-rs/server/src/child_reaper.rs:164]: Basically, I am not sure if `WaitStatus::StillAlive` covers stopped and continued process, and if not, then whether we should break here for such or continue waiting? @saschagrunert, thoughts?

Comment (kwilczynski) [reply conmon-rs/server/src/child_reaper.rs:164]: Makes sense. Thank you!

**Pull request merged**